rules/dispatch.go

```diff
@@ -102,112 +113,69 @@ func (r *DefaultRuleRenderer) hostDispatchChains(
 			ChainDispatchFromHostEndpoint,
 			ChainDispatchToHostEndpoint,
 			false,
 			false,
-			false,
 		)
-
 	}
 
 	return append(
-		r.dispatchChains(
+		r.dispatchEndpointChains(
 			names,
-			epMarkMapper,
 			HostFromEndpointPfx,
 			HostToEndpointPfx,
 			ChainDispatchFromHostEndpoint,
 			ChainDispatchToHostEndpoint,
 			false,
 			false,
-			false,
 		),
-		r.dispatchChains(
+		r.dispatchEndpointChains(
 			names,
-			epMarkMapper,
 			HostFromEndpointForwardPfx,
 			HostToEndpointForwardPfx,
 			ChainDispatchFromHostEndPointForward,
 			ChainDispatchToHostEndpointForward,
 			false,
 			false,
-			false,
 		)...,
 	)
 }
 
-func (r *DefaultRuleRenderer) dispatchChains(
+func (r *DefaultRuleRenderer) dispatchEndpointChains(
 	names []string,
-	epMarkMapper EndpointMarkMapper,
 	fromEndpointPfx,
 	toEndpointPfx,
 	dispatchFromEndpointChainName,
 	dispatchToEndpointChainName string,
 	dropAtEndOfFromChain bool,
 	dropAtEndOfToChain bool,
-	useEndPointMark bool,
 ) []*Chain {
 	// Sort interface names so that rules in the dispatch chain are ordered deterministically.
 	// Otherwise we would reprogram the dispatch chain when there is no real change.
 	sort.Strings(names)
-	log.WithField("ifaceNames", names).Debug("Rendering dispatch chains")
+	log.WithField("ifaceNames", names).Debug("Rendering endpoint dispatch chains")
 
 	// Since there can be >100 endpoints, putting them in a single list adds some latency to
 	// endpoints that are later in the chain.  To reduce that impact, we build a shallow tree of
 	// chains based on the prefixes of the chains.
-
-	// Start by figuring out the common prefix of the endpoint names.  Commonly, this will
-	// be the interface prefix, e.g. "cali", but we may get lucky if multiple interfaces share
-	// a longer prefix.
-	commonPrefix := stringutils.CommonPrefix(names)
-	log.WithField("commonPrefix", commonPrefix).Debug("Calculated common prefix")
-
-	// Then, divide the names into bins based on their next character.
-	prefixes := []string{}
-	prefixToNames := map[string][]string{}
-	lastName := ""
-	for _, name := range names {
-		if name == "" {
-			log.Panic("Unable to render dispatch chain. Empty interface name.")
-		}
-		if name == lastName {
-			log.WithField("ifaceName", name).Error(
-				"Multiple endpoints with same interface name detected. " +
-					"Incorrect policy may be applied.")
-			continue
-		}
-		prefix := commonPrefix
-		if len(name) > len(commonPrefix) {
-			prefix = name[:len(commonPrefix)+1]
-		}
-		if _, present := prefixToNames[prefix]; !present {
-			// Record the prefixes in sorted order (if we iterate over the map, we get a
-			// random order, which we don't want).
-			prefixes = append(prefixes, prefix)
-		}
-		prefixToNames[prefix] = append(prefixToNames[prefix], name)
-		lastName = name
-	}
+	commonPrefix, prefixes, prefixToNames := r.divideEndpointNamesToPrefixTree(names)
 
 	rootFromEndpointRules := make([]Rule, 0)
 	rootToEndpointRules := make([]Rule, 0)
```

Comment:
Comment on Line 185 now out of date since this function no longer deals with the mark chains.

Reply:
Done.